### PR TITLE
Add a partial to download the assets from CDNs 

### PR DESCRIPTION
**Pull request by zecakeh**
Instead of downloading them manually when upgrading the theme.

This should be reviewed by commit, the first one resets the modified files to their version upstream so the changes in the second commit are cleaner.

### CODE REVIEW

**Comment by zecakeh** [comment]
To see the differences with upstream: https://github.com/google/docsy/compare/v0.8.0...zecakeh:matrix-org-docsy:cdn-shortcode.

**Comment by zecakeh** [comment]
Another nice thing that this adds is that the assets will only be bundled if we need them. So currently we only download jQuery with the spec.

File: layouts/partials/dl-asset.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  A partial to download an asset from a CDN at build time and serve it from our
+  server. It outputs HTML attributes of the form:
+
+  {src/href}="{permalink}" integrity="{hash}"
+
+  Parameters:
+
+  * `src` or `href`: a string, the URL of the asset to download. The name of the
+    parameter will be reused as the attribute for the permalink.
+  * `integrity`: a boolean, whether to output the integrity hash.
```
Comment (richvdh) [inline layouts/partials/dl-asset.html]:
```suggestion
  * `integrity`: a boolean, whether to output the integrity attribute. If unspecified, defaults to `false`.
```

File: layouts/partials/dl-asset.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  {{ $url = .href -}}
+  {{ $attribute = "href" -}}
+{{ end -}}
+
+{{ with resources.GetRemote $url -}}
+  {{ with .Err -}}
+    {{ errorf "Could not retrieve asset at '%s' from CDN. Reason: %s." $url . -}}
+  {{ else -}}
+    {{ $attribute }}="{{ .RelPermalink -}}"
+    {{- with (and $integrity .Data.Integrity) }} integrity="{{ . }}" {{ end -}}
```
Comment (richvdh) [inline layouts/partials/dl-asset.html]: does this rely on `resources.GetRemote` to support `.Data.Integrity` in the result? Does it? It doesn't seem to be documented at https://gohugo.io/methods/resource/data/?

Comment (zecakeh) [reply layouts/partials/dl-asset.html]: Oh your right, I forgot to pass it through [`fingerprint`](https://gohugo.io/functions/resources/fingerprint/).

Comment (richvdh) [reply layouts/partials/dl-asset.html]: I have to say, I wonder if it's worth the effort. It looks like it's only used for one resource, which doesn't currently have an integrity check.

Comment (richvdh) [reply layouts/partials/dl-asset.html]: (we could continue to omit the integrity check for that resource, or alternatively hardcode a specific version and hash)

Comment (zecakeh) [reply layouts/partials/dl-asset.html]: I don't have strong feelings about this, @anoadragon453 was the one that suggested to generate the hash automatically in the Matrix Spec & Docs Authoring room.

Comment (anoadragon453) [reply layouts/partials/dl-asset.html]: Generating the hash is preferable as doing it manually adds extra steps, but if it's extra fiddly in the code then we can opt to not bother.

Integrity checks prevent against the upstream resource subtly changing a file based on who's downloading it. I don't think that is a realistic threat within the spec website's threat model though. So if it ends up being a pain, let's just drop it.

Comment (richvdh) [reply layouts/partials/dl-asset.html]: We've got this far without an integrity check...

> Generating the hash is preferable as doing it manually adds extra steps,

yes, though what are the actual ramifications if we end up on an old version of `markmap-autoloader`? (WTF is a markmap useful for anyway?)

Comment (anoadragon453) [reply layouts/partials/dl-asset.html]: We've had integrity checks on the assets so far, see [this commit](https://github.com/google/docsy/commit/a0032f8db919a6c67ba6cdef2c455f105b6272a2).

I think the issue is less an old version of a library, but more if the CDN somehow serves malicious JS. But I suppose there aren't any credentials on spec.matrix.org, so the attack surface is probably limited.

Comment (anoadragon453) [reply layouts/partials/dl-asset.html]: From internal discussion, I realised that placing an integrity check doesn't buy you much when the hash for the check is being served by the same webserver that's serving the files you want to check. That same server could just modify the hash when it serves you the HTML.

While it made sense when we were pulling the libraries from different CDNs, it doesn't make sense when we're vendoring the libraries ourselves.

We instead just have TOFU from the time we build the theme and vendor the libraries, which I think is fine for our threat model.

TL;DR @zecakeh I believe we can remove the `integrity` tag on any js/css assets that we're serving ourselves. They don't buy us any extra security.

Comment (zecakeh) [reply layouts/partials/dl-asset.html]: Alright I simplified the partial to only take a URL and return the relative path to the asset.

File: layouts/partials/dl-asset.html (hunk 1 above)
Comment (zecakeh) [reply layouts/partials/dl-asset.html]: The parameter has been removed, so this does not apply anymore.

**Review by richvdh — APPROVED** [review]
right, this is much cleaner now!
